### PR TITLE
Added Advanced section that include resource type development topics.

### PR DESCRIPTION
**Pull request by mrinaudo-aws**
*Issue #, if available:*

*Description of changes:* Added Advanced section with resource type development-related topics.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by jaymccon — DISMISSED** [review]
lgtm

File: workshop/content/advanced/resource-types/prerequisites/_index.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+---
+title: 'Prerequisites'
+date: 2021-11-16T20:50:53Z
+weight: 320
+---
+
+### Resource Type Development Tools
```
Comment (rezabekf) [inline workshop/content/advanced/resource-types/prerequisites/_index.md:7]: Can we provide example how to install pre-req for maybe one OS. It should be more visible what pre-reqs are needed, maybe as a list with links? Right now there is a lot of text and requirements can be overlooked imo. Some pre-req examples are [here](https://cfn101.workshop.aws/prerequisites.html)

File: workshop/content/advanced/resource-types/prerequisites/_index.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+* for the [resource type example in Python](https://github.com/aws-cloudformation/aws-cloudformation-samples/tree/main/resource-types/awssamples-ec2-importkeypair/python) described in the next lab, follow [notes](https://github.com/aws-cloudformation/aws-cloudformation-samples/tree/main/resource-types/awssamples-ec2-importkeypair/python#unit-tests) to install, with `pip`, the `pytest-cov` and `cloudformation-cli-python-lib` packages that are needed for running unit tests for the sample application.
```
Comment (rezabekf) [inline workshop/content/advanced/resource-types/prerequisites/_index.md:13]: Can we provide the `pip` commands what packages to install here?

File: workshop/content/advanced/resource-types/prerequisites/_index.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+* follow [steps](https://docs.aws.amazon.com/cloudformation-cli/latest/userguide/what-is-cloudformation-cli.html#resource-type-setup) to install the [CloudFormation Command Line Interface (CLI)](https://docs.aws.amazon.com/cloudformation-cli/latest/userguide/what-is-cloudformation-cli.html): this includes installing/checking the version of Python installed on your machine, the installation of the CloudFormation CLI and language plugins. You will use the CloudFormation Command Line Interface for operations that include generating a new project for your resource type with source code stubs, validating your resource type specification against the [resource type schema](https://docs.aws.amazon.com/cloudformation-cli/latest/userguide/resource-type-schema.html), running contract tests and submitting your resource type to the CloudFormation registry;
```
Comment (rezabekf) [inline workshop/content/advanced/resource-types/prerequisites/_index.md:12]: Can we clarify that first we install version 1.0 and upgrade to version 2.0 once inside the project? The documentation doesn't go into much detail here.

File: workshop/content/advanced/resource-types/key-concepts/_index.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
+---
+title: 'Key Concepts'
+date: 2021-11-16T20:44:31Z
+weight: 310
+---
+
+### Getting started
```
Comment (rezabekf) [inline workshop/content/advanced/resource-types/key-concepts/_index.md:7]: Can we move this section onto `workshop/content/advanced/resource-types/_index.md` Its rather short and I feel it doesn't need own page. After introduction it may be nice to have a key concepts on the same page.

Comment (mrinaudo-aws) [reply workshop/content/advanced/resource-types/key-concepts/_index.md:7]: Sure - thank you!  Will do!

Comment (mrinaudo-aws) [reply workshop/content/advanced/resource-types/key-concepts/_index.md]: Done

File: workshop/content/advanced/resource-types/prerequisites/_index.md (hunk 1 above)
Comment (mrinaudo-aws) [reply workshop/content/advanced/resource-types/prerequisites/_index.md:7]: Thank you - I will add a list with links for the AWS SAM CLI.

File: workshop/content/advanced/resource-types/prerequisites/_index.md (hunk 3 above)
Comment (mrinaudo-aws) [reply workshop/content/advanced/resource-types/prerequisites/_index.md:12]: The example in the documentation is for the 1.0 to 2.0 upgrade if version 1.0 is installed. Will add an example and will reference the documentation reference - thank you!

File: workshop/content/advanced/resource-types/prerequisites/_index.md (hunk 2 above)
Comment (mrinaudo-aws) [reply workshop/content/advanced/resource-types/prerequisites/_index.md:13]: Yes - will do!

**Review by rezabekf — DISMISSED** [review]

File: workshop/content/advanced/resource-types/example-in_Python/_index.md [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+* `src`: contains a directory named after your resource type name, inside of which you should find the `models.py` file, that you manage with the CloudFormation CLI when you make schema changes, and the `handlers.py` file, that is where you will start adding your code for the CRUDL implementation logic. Open the `src/handlers.py` file with a text editor of your choice, and **familiarize with the handlers' structure** described in `create_handler`, `update_handler`, `delete_handler`, `read_handler`, `list_handler` functions;
+* `resource-role.yaml`: file created by the CloudFormation CLI, that describes an [AWS Identity and Access Management](https://aws.amazon.com/iam/) (IAM) role whose `PolicyDocument` contains permissions you indicate in the schema for your handlers in the `handlers` section.  CloudFormation assumes this role to manage resources on your behalf as part of CRUDL operations;
+* `template.yml`: [AWS Serverless Application Model](https://aws.amazon.com/serverless/sam/) (SAM) template used as part of resource type testing.
+
+
+
+#### Example resource type walkthrough
```
[the remark this answers is not in the document]
Comment (mrinaudo-aws) [reply workshop/content/advanced/resource-types/example-in_Python/_index.md:27]: The lab mainly focuses on thought process and example considerations to make when designing a resource model and building handler logic. It offers the existing resource as a sample to compare takeaways, and to see those applied in the sample itself. It then also provides hands-on actions on topics that include creating the infrastructure for contract tests, running unit tests, submitting to the registry and offering, in the challenge section, the action of writing a JSON file test for manually testing a handler.

[the remark this answers is not in the document]
Comment (stilvoid) [reply workshop/content/advanced/resource-types/example-in_Python/_index.md:27]: In the context of a hands-on, real-world workshop with 100 users in a room expecting to learn how to build resource types, I feel it would be good to have some step-by-step guidance on how to get started from scratch. We could include this advice in addition to providing an example resource type to work through.

File: workshop/content/advanced/resource-types/example-in_Python/_index.md [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+
+
+### Start Lab
+
+In this lab, you will go on a deep dive for the code of an existing, example resource type: you will explore steps and a number of considerations to make when modeling and implementing a resource type. For information on how to get started with the CloudFormation CLI to create a new project, see [Walkthrough: Develop a resource type](https://docs.aws.amazon.com/cloudformation-cli/latest/userguide/resource-type-walkthrough.html).
+
+
+#### Example resource type walkthrough
+
+Let's dive deeper into key considerations you should make when you design the model schema of your resource type. As an example, you will use an existing, sample resource type called [AWSSamples::EC2::ImportKeyPair](https://github.com/aws-cloudformation/aws-cloudformation-samples/tree/main/resource-types/awssamples-ec2-importkeypair/python), that illustrates an example of importing and managing an imported [Amazon EC2](https://aws.amazon.com/ec2/) key pair with CloudFormation.
```
Comment (stilvoid) [inline workshop/content/advanced/resource-types/example-in_Python/_index.md:29]: Given the example repo is outside of the workshop repo, there's a risk that it could move or change. Also, given we provide sample code in other parts of this workshop under the `code` directory, it might be best to vendor this example resource type there too.

File: workshop/content/advanced/resource-types/example-in_Python/_index.md [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+* `awssamples-ec2-importkeypair.json`: this file, named after the resource type name choice, contains the schema for the resource. When you develop a resource type, you use this file to **describe the model for the resource**;
+* `src`: contains a directory named after the resource type name, inside of which you should find:
+  - the `models.py` file, that is managed by the CloudFormation CLI on your behalf when you make schema changes, and
+  - the `handlers.py` file, that is where the resource type developer adds code for the CRUDL implementation logic. Open the `src/handlers.py` file with a text editor of your choice, and **familiarize with the handlers' structure** described in `create_handler`, `update_handler`, `delete_handler`, `read_handler`, `list_handler` functions;
+* `resource-role.yaml`: file managed by the CloudFormation CLI, that describes an [AWS Identity and Access Management](https://aws.amazon.com/iam/) (IAM) role whose `PolicyDocument` contains permissions the resource type developer indicates in the `handlers` section of the schema file.  CloudFormation assumes this role to manage resources on behalf of the user as part of CRUDL operations;
```
Comment (stilvoid) [inline workshop/content/advanced/resource-types/example-in_Python/_index.md:53]: Although it is implied, we should explicitly state that the role here is used to manage resources _for this resource type_

File: workshop/content/advanced/resource-types/example-in_Python/_index.md [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+* `template.yml`: [AWS Serverless Application Model](https://aws.amazon.com/serverless/sam/) (SAM) template used as part of resource type testing.
+
+In the next section, you will explore aspects of the thought process of *implementing a resource type by starting from the model*, and you will compare such steps with example code content you just chose to clone or download on your machine.
+
+
+#### Resource modeling
+
+The first step in creating a resource type is to define a schema that describes properties for your resource, as well as permissions needed for CloudFormation to manage the resource on your behalf.
```
Comment (stilvoid) [inline workshop/content/advanced/resource-types/example-in_Python/_index.md:61]: This whole section would be _a lot_ clearer by including some of the schema in the workshop text.

Comment (stilvoid) [reply workshop/content/advanced/resource-types/example-in_Python/_index.md:61]: This section makes no mention of JSON Schema. Should definitely include a link to https://json-schema.org/ where users can brush up on that first.

File: workshop/content/advanced/resource-types/_index.md [hunk 9]
```diff
@@ -0,0 +1,36 @@
+---
+title: 'Resource Types'
+date: 2021-11-16T16:51:46Z
+weight: 300
+---
+
+### Introduction
+
+This part of the workshop focuses on how you can extend creation, provisioning, and management capabilities of [AWS CloudFormation](https://aws.amazon.com/cloudformation/) with resource types you develop.
```
[the remark this answers is not in the document]
Comment (stilvoid) [reply workshop/content/advanced/resource-types/_index.md:9]: After reading further down, I'm doubling down on this comment. A previous section introducting modules as a pre-req to resource types would simplify this part of the workshop _a lot_.

**Review by stilvoid — COMMENTED** [review]
The content is good but overall it feels a bit wordy with a much higher paragraph-to-code-block ratio than the other parts of the cfn101 workshop. A few code snippets to highlight the points throughout would improve clarity.
